src/wire/dhcp.rs

```diff
@@ -0,0 +1,629 @@
+
+        let mut options = packet.options()?;
+        while options.len() > 0 {
+            let (next_options, option) = DhcpOption::parse(options)?;
+            match option {
+                DhcpOption::EndOfList => break,
+                DhcpOption::Pad => {},
+                DhcpOption::MessageType(value) => {
+                    if value.opcode() == packet.opcode() {
+                        message_type = Ok(value);
+                    }
+                },
+                DhcpOption::RequestedIp(ip) => {
+                    requested_ip = Some(ip);
+                }
+                DhcpOption::ServerIdentifier(ip) => {
+                    server_identifier = Some(ip);
+                }
+                DhcpOption::Other {kind: field::OPT_PARAMETER_REQUEST_LIST, data} => {
+                    parameter_request_list = Some(data);
+                }
+                DhcpOption::Other {..} => {}
+            }
+            options = next_options;
+        }
+
+        let broadcast = packet.broadcast_flag();
+
+        Ok(Repr {
+            transaction_id, client_hardware_address, client_ip, your_ip, server_ip, relay_agent_ip,
+            broadcast, requested_ip, server_identifier, parameter_request_list,
+            message_type: message_type?,
+        })
+    }
+
+    /// Emit a high-level representation into a Transmission Control Protocol packet.
+    pub fn emit<T>(&self, packet: &mut Packet<&mut T>) -> Result<()>
+            where T: AsRef<[u8]> + AsMut<[u8]> + ?Sized {
+        packet.set_opcode(self.message_type.opcode());
+        packet.set_hardware_type(Hardware::Ethernet);
+        packet.set_hardware_len(6);
+        packet.set_transaction_id(self.transaction_id);
+        packet.set_client_hardware_address(self.client_hardware_address);
+        packet.set_hops(0);
+        packet.set_secs(0); // TODO
+        packet.set_magic_number(0x63825363);
+        packet.set_client_ip(self.client_ip);
+        packet.set_your_ip(self.your_ip);
+        packet.set_server_ip(self.server_ip);
+        packet.set_relay_agent_ip(self.relay_agent_ip);
+        packet.set_broadcast_flag(self.broadcast);
+
+        {
+            let mut options = packet.options_mut()?;
+            let tmp = options; options = DhcpOption::MessageType(self.message_type).emit(tmp);
+            if let Some(list) = self.parameter_request_list {
+                let option = DhcpOption::Other{ kind: field::OPT_PARAMETER_REQUEST_LIST, data: list };
+                let tmp = options; options = option.emit(tmp);
+            }
+            if let Some(ip) = self.server_identifier {
+                let tmp = options; options = DhcpOption::ServerIdentifier(ip).emit(tmp);
+            }
+            if let Some(ip) = self.requested_ip {
+                let tmp = options; options = DhcpOption::RequestedIp(ip).emit(tmp);
+            }
+            DhcpOption::EndOfList.emit(options);
+        }
+
+        Ok(())
+    }
+}
```

Comment:
The other `Repr` and `Packet` types in `wire` have tests. It might be good to add some basic unit tests.